accepted/00065-modularisation.md

```diff
@@ -0,0 +1,90 @@
+- Make batch operations the default and single target a special case. This is so implementations can be done with scalability in mind.
+- Merge related existing implementations across the codebase to implement the interface
+- Make sure the interface does not imply a specific implementation (i.e don't reference salt if all you do is gathering information from a system.)
+- Refactor other components to take an implementation of this interface as an argument instead of reaching into global space.
+- write new and reuse existing tests for the interface
+- move interface with related datatypes and implementations into package with fitting name for the component.
+
+### Interface guidelines?
+
+- It should be focused on a particular area and not mix different topics.
+- It should be minimal meaning only the minimal set of methods needed by our business logic.
+- It should give immediately useful information and not burden the user to do much additional work to get to the useful information. (i.e don't return plain json and let business logic parse it).
+- it shoud be precise in its types and minimize invalid usage.
+- Don't be scared to define multiple interface to separate areas properly and let them build ontop of each other.
+
+
+## Bootstrapping
+
+Bootstrapping is about creating our instances of our different "services" and assembling them to then pass on into our business logic.
+The assembly should happen at the very start and then everything below that will only get things passed down but not instanciate any services themselves.
+The Start in this case is kind of multiple places tomcat and taskomatic are differnt processes and have there own entry points so both need to have some bootstrapping code.
+In tomcat the situation is maybe a little bit more complicated since our business logic lives below filters which are instanciated by tomcat logic based on xml configuration.
+
+- One way to go could be looking into programatic initialisation and registration of filters.
+- Use some limited shared global state just to pass the initial instances between entry points.
+- in the worst case duplicating instances per filter so spark and xmlrpc would not share the same instances.
+
+
+
+## Minimal example
+
+This ia a minimal example extracting most of the interactions with salt into an interface. Its not complete
+as it lacks more refinement i.e getting rid of JsonElement in result or splitting it up into more topical interface i.e having libvirt interaction in its own interface. But it should give a general idea of how this rfc looks applied to our code.
+
+[salt-branch](https://github.com/uyuni-project/uyuni/compare/salt-interface?expand=1)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Refactorings in areas that are not well understood by anyone may be hard and lead to short term breakage.
+- measures will have to be found to ensure established interfaces will continue to be respected.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
```

Comment:
How would the traditional (legacy) stack fit into this picture? Would it be still there in your imagined scenario?

Reply:
@mbologna I guess, something like that:

```java
interface InterfaceName {
    
    // declare constant fields
    // declare methods that abstract 
    // by default.
}
```

😉 

Imagine you want to un-install that traditional thing away. So you would `zypper rm uyuni-traditional-connector` or something like that. But before that, you probably would like to crawl across the whole codebase, rip off all traditional-related bare methods and random classes, then place them behind some common interface, so they became an implementation detail.

Which is the same as the rest of Kevin's proposal. So I don't see it as an "unresolved question". But let's re-ask this tomorrow at his session maybe?

Reply:
> Imagine you want to un-install that traditional thing away. So you would `zypper rm uyuni-traditional-connector` or something like that. But before that, you probably would like to crawl across the whole codebase, rip off all traditional-related bare methods and random classes, then place them behind some common interface, so they became an implementation detail.
> 
> Which is the same as the rest of Kevin's proposal. So I don't see it as an "unresolved question". But let's re-ask this tomorrow at his session maybe?

Is it still unresolved: while doing the refactoring, should we keep the traditional stack or should we drop it? It sounds like, from your proposal, that you want to keep it. But I haven't got Kevin's POV from the RFC text.

Reply:
> should we keep the traditional stack or should we drop it?

I am not sure if this is relevant to the current RFC.